### PR TITLE
Added consistent hashing functions

### PR DESCRIPTION
**Pull request by ludv1x**

### CODE REVIEW

File: dbms/src/Functions/FunctionsConsistentHashing.h [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+        Field buckets_field = (*buckets_col)[0];
+        if (buckets_field.getType() == Field::Types::Int64)
+            check_range(buckets_field.safeGet<Int64>());
+        else if (buckets_field.getType() == Field::Types::UInt64)
+            check_range(buckets_field.safeGet<UInt64>());
+        else
+            throw Exception("Illegal type " + String(buckets_field.getTypeName()) + " of the second argument of function " + getName(),
+                            ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+
```
Comment (alexey-milovidov) [inline dbms/src/Functions/FunctionsConsistentHashing.h]: Excessive whitespace.

File: libs/yandex-consistent-hashing/yandex/bitops.h [hunk 2]
```diff
@@ -0,0 +1,55 @@
+#pragma once
+#include <stdint.h>
+#include <limits>
+#include <type_traits>
+
+// Assume little endian
+
+inline uint16_t & LO_16(uint32_t & x) { return reinterpret_cast<uint16_t *>(&x)[0]; }
```
Comment (alexey-milovidov) [inline libs/yandex-consistent-hashing/yandex/bitops.h]: This breaks strict aliasing.

File: libs/yandex-consistent-hashing/yandex/bitops.h [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+inline uint16_t & HI_16(uint32_t & x) { return reinterpret_cast<uint16_t *>(&x)[1]; }
+
+inline uint32_t & LO_32(uint64_t & x) { return reinterpret_cast<uint32_t *>(&x)[0]; }
+inline uint32_t & HI_32(uint64_t & x) { return reinterpret_cast<uint32_t *>(&x)[1]; }
+
+
+#if defined(__GNUC__)
```
Comment (alexey-milovidov) [inline libs/yandex-consistent-hashing/yandex/bitops.h]: Better to add a comment that clang also defines __GNUC__

File: libs/yandex-consistent-hashing/yandex/bitops.h [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+            // Y_ASSERT(value); // because __builtin_clz* have undefined result for zero.
+            return std::numeric_limits<unsigned long>::digits - __builtin_clzl(value);
+        }
+
+        inline unsigned GetValueBitCountImpl(unsigned long long value) noexcept {
+            // Y_ASSERT(value); // because __builtin_clz* have undefined result for zero.
+            return std::numeric_limits<unsigned long long>::digits - __builtin_clzll(value);
+        }
+#else
+        /// Stupid realization for non-GCC. Can use BSR from x86 instructions set.
```
Comment (alexey-milovidov) [inline libs/yandex-consistent-hashing/yandex/bitops.h]: Misleading comment about GCC.

File: libs/yandex-consistent-hashing/yandex/bitops.h [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        template <typename T>
+        inline unsigned GetValueBitCountImpl(T value) noexcept {
+            // Y_ASSERT(value);     // because __builtin_clz* have undefined result for zero.
+            unsigned result = 1; // result == 0 - impossible value, see Y_ASSERT().
```
Comment (alexey-milovidov) [inline libs/yandex-consistent-hashing/yandex/bitops.h]: Better to totally remove this code.

File: libs/yandex-consistent-hashing/yandex/popcount.h [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+static inline uint32_t PopCountImpl(uint64_t n) {
+#if defined(_MSC_VER) && !defined(_i386_)
+    return __popcnt64(n);
+#else
+#if defined(__x86_64__)
+
+    if (HavePOPCNTInstr) {
+        uint64_t r;
+
+        __asm__("popcnt %1, %0;"
```
Comment (alexey-milovidov) [inline libs/yandex-consistent-hashing/yandex/popcount.h]: Ok, but there is `__builtin_popcountll` in gcc and clang.

Comment (alexey-milovidov) [reply libs/yandex-consistent-hashing/yandex/popcount.h]: The advantage is that also may work efficiently on ARM.

File: dbms/src/Functions/registerFunctions.cpp [hunk 7]
```diff
@@ -60,6 +61,7 @@ void registerFunctions()
     registerFunctionsExternalModels(factory);
     registerFunctionsFormatting(factory);
     registerFunctionsHashing(factory);
+    registerFunctionsConsistentHashing(factory);
```
Comment (alexey-milovidov) [inline dbms/src/Functions/registerFunctions.cpp]: I would also like to see performance test.

File: dbms/src/Functions/FunctionsConsistentHashing.h [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+
+        const auto & hash_col_source = block.getByPosition(arguments[0]).column;
+        ColumnPtr hash_col = (hash_col_source->isColumnConst()) ? hash_col_source->convertToFullColumnIfConst() : hash_col_source;
```
Comment (alexey-milovidov) [inline dbms/src/Functions/FunctionsConsistentHashing.h]: No need for this. Look at `useDefaultImplementationForConstants` and `getArgumentsThatAreAlwaysConstant`.

Comment (alexey-milovidov) [reply dbms/src/Functions/FunctionsConsistentHashing.h]: (The function should be constexpr for constant arguments.)

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
No mention about the author of Yandex function.

File: dbms/src/Functions/FunctionsConsistentHashing.cpp [hunk 9]
```diff
@@ -0,0 +1,14 @@
+#include "FunctionsConsistentHashing.h"
+#include <Functions/FunctionFactory.h>
+
+
+namespace DB
+{
+
+void registerFunctionsConsistentHashing(FunctionFactory & factory)
```
Comment (zhang2014) [inline dbms/src/Functions/FunctionsConsistentHashing.cpp]: It's weird as a function :smile:

**Pull request merged**